obi/ui/views.py

```diff
@@ -395,8 +397,57 @@ def _summon_query(request, scope='all'):
             match['publicationplace'] = document['PublicationPlace'][0]
         if document.get('hasFullText', []):
             match['hasFullText'] = document['hasFullText']
+        else:
+            match['hasFullText'] = False
         if document.get('Institution', []):
             match['institution'] = document['Institution'][0]
+
+        # Start with the result provided by Summon,
+        # but overwrite if we can do better
+        match['url'] = document['link']
+        if scope == 'articles' and match['hasFullText'] is True:
+            # Check whether Summon-provided link resolves to
+            # findit.library.gwu.edu.
+            # If it takes us elsewhere (to the resource), use that.
+            # But if it takes us to findit, attempt to resolve w/openURL.
+
+            # Send a HEAD request, not a GET request, to avoid the
+            # added time involved to provide the message body (which
+            # in this case could be an entire scholarly article)
+            r = requests.head(document['link'],
+                              timeout=settings.SUMMON_TIMEOUT_SECONDS)
+            if 200 <= r.status_code < 400:
+                resolved_url = r.headers['Location']
+                if not(resolved_url.startswith('http://findit.library.gwu') or
+                       resolved_url.startswith('https://findit.library.gwu')):
+                    # Summon already will resolve to the resource;
+                    # great, use that!
+                    match['url'] = resolved_url
+                else:
+                    # Construct the OpenURL request and get the OpenURL XML document
+                    xmlurl = 'http://uz4ug4lz9g.openurl.xml.serialssolutions.com/openurlxml?version=1.0&' + document['openUrl']
+                    f = urllib2.urlopen(xmlurl)
+                    tree = f.read()
+                    f.close()
+                    data = xmltodict.parse(tree)
+                    match['OpenUrl'] = xmlurl
+
+                    # Traverse the tree and look for the node(s) containing
+                    # article link, if present
+                    try:
+                        urlnodes = data['ssopenurl:openURLResponse']['ssopenurl:results']['ssopenurl:result']['ssopenurl:linkGroups']['ssopenurl:linkGroup']['ssopenurl:url']
```

Comment:
It looks like this will only look at the first linkGroup, not all of them. In some cases, the first linkGroup is not one that will have a url with type "article". (It could  be a linkGroup with type "termsOfUse" instead of "holding" and the url a link to terms of use. In other cases, the first linkGroup does have type "holdings" but it's not online holdings, it's print holdings and has no url with type "article".)

Example search: manchester by the sea

Example openURL XML (currently, the third result from the example):
http://uz4ug4lz9g.openurl.xml.serialssolutions.com/openurlxml?version=1.0&ctx_ver=Z39.88-2004&ctx_enc=info%3Aofi%2Fenc%3AUTF-8&rfr_id=info%3Asid%2Fsummon.serialssolutions.com&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&rft.genre=article&rft.atitle=Health%2C+Medicine%2C+and+the+Sea%3A+Australian+Voyages%2C+c.1815-60.+By+Katherine+Foxhall.+Manchester+University+Press.+2012.+xiii+%2B+250pp.+%C2%A360.00&rft.jtitle=History&rft.au=Lamb%2C+Jonathan&rft.date=2014-01-01&rft.issn=0018-2648&rft.eissn=1468-229X&rft.volume=99&rft.issue=334&rft.spage=186&rft.epage=187&rft_id=info:doi/10.1111%2F1468-229X.12049_37&rft.externalDBID=n%2Fa&rft.externalDocID=10_1111_1468_229X_12049_37